lib/actions/modemTargetActions.js

```diff
@@ -134,3 +136,120 @@ export const performUpdate = () => (dispatch, getState) => {
+        } catch (e) {
+            dispatch(processUpdateAction(
+                progress.process,
+                totalPercentage,
+                totalDuration,
+                step,
+            ));
+            return;
+        }
+
+        if (dfuProcess && dfuProcess.operation === 'upload_image') {
+            totalPercentage = dfuProcess.progress_percentage;
+            totalDuration = dfuProcess.duration;
+
+            progressMsg = dfuProcess.message || progressMsg;
+            dispatch(processUpdateAction(
+                progressMsg,
+                totalPercentage,
+                totalDuration,
+                step,
+            ));
+        }
+    };
+
+    const modemCallback = err => {
+        if (err) {
+            logger.error(`Modem DFU failed with error: ${err}`);
+            let errorMsg = err.message || err;
+            if (err.message.includes('0x4')) {
+                errorMsg = 'Failed with internal error. '
+                    + 'Please click Erase all button and try updating modem again.';
+            }
+            dispatch(writingFailAction(errorMsg));
+            return;
+        }
+        logger.info('Modem DFU completed successfully!');
+        dispatch(writingSucceedAction());
+    };
+
+    const mcuCallback = err => {
+        if (err) {
+            let errorMsg = err.message || err;
+
+            // Program without setting device in MCUboot mode will throw such an error:
+            // Errorcode: CouldNotCallFunction (0x9)
+            // Lowlevel error: Unknown value (ffffff24)
+            // Interpret the error in a way that user can understand it easily.
+            if (errorMsg.includes('0x9')) {
+                errorMsg = 'Please make sure that the device is in MCUboot mode and try again.';
+            }
+
+            logger.error(`MCUboot DFU failed. ${errorMsg}`);
+            dispatch(writingFailAction(errorMsg));
+            return;
+        }
+        const modemUartDfu = () => nrfjprog.programModemUartDFU(
+            noSerialNumber,
+            fileName,
+            port2,
+            baudrate1m,
+            timeout,
+            progressCallback,
+            modemCallback,
+        );
+        step += 1;
+        setTimeout(modemUartDfu, 2000);
+    };
+
+    nrfjprog.programMcuBootDFU(
+        noSerialNumber,
+        mcubootFilePath,
+        port,
+        baudrate115k2,
+        timeout,
+        progressCallback,
+        mcuCallback,
+    );
+};
+
+export const performUpdate = () => (dispatch, getState) => {
+    dispatch(writingStartAction());
+    dispatch(processUpdateAction(MODEM_DFU_STARTING));
+    logger.info('Modem DFU starts to write...');
+
+    const { modemFwName: fileName } = getState().app.modem;
+    const serialNumber = Number(getState().app.target.serialNumber);
+    logger.info(`Writing to device with serial number: ${serialNumber}`);
+    logger.info(`Writing to device with file: ${fileName}`);
```

Comment:
I suggest to have simplify the two lines by fixing the grammar:
`Writing ${fileName} to device ${serialNumber}`